addons/mail/static/src/new/messaging_hook.js

```diff
@@ -36,3 +37,117 @@ export function useMessageHighlight(duration = 2000) {
+            upload("/mail/attachment/upload", [file], {
+                buildFormData(formData) {
+                    formData.append("thread_id", thread.resId || thread.id);
+                    formData.append("thread_model", thread.resModel || "mail.channel");
+                    formData.append("is_pending", true);
+                    formData.append("temporary_id", tmpId);
+                },
+            }).catch((e) => {
+                if (e.name !== "AbortError") {
+                    throw e;
+                }
+            });
+        },
+        async unlink(attachment) {
+            const abort = abortByUploadId[attachment.id];
+            if (abort) {
+                abort();
+                return;
+            }
+            await messaging.unlinkAttachment(attachment);
+            removeFromArrayWithPredicate(state.attachments, ({ id }) => id === attachment.id);
+        },
+        async unlinkAll() {
+            const proms = [];
+            this.attachments.forEach((attachment) => proms.push(this.unlink(attachment)));
+            await Promise.all(proms);
+            this.reset();
+        },
+        reset() {
+            abortByUploadId = {};
+            uploadingAttachmentIds.clear();
+            state.attachments = [];
+        },
+    });
+    useBus(bus, "FILE_UPLOAD_ADDED", ({ detail: { upload } }) => {
+        if (!uploadingAttachmentIds.has(parseInt(upload.data.get("temporary_id")))) {
+            return;
+        }
+        const threadId = upload.data.get("thread_id");
+        const threadModel = upload.data.get("thread_model");
+        const originThread =
+            messaging.state.threads[
+                threadModel === "mail.channel" ? parseInt(threadId) : `${threadModel},${threadId}`
+            ];
+        abortByUploadId[upload.id] = upload.xhr.abort.bind(upload.xhr);
+        state.attachments.push({
+            filename: upload.title,
+            id: upload.id,
+            mimetype: upload.type,
+            name: upload.title,
+            originThread,
+            size: upload.total,
+            uploading: true,
+        });
+    });
+    useBus(bus, "FILE_UPLOAD_LOADED", ({ detail: { upload } }) => {
+        const tmpId = parseInt(upload.data.get("temporary_id"));
+        if (!uploadingAttachmentIds.has(tmpId)) {
+            return;
+        }
+        uploadingAttachmentIds.delete(tmpId);
+        delete abortByUploadId[upload.id];
+        const response = JSON.parse(upload.xhr.response);
+        if (response.error) {
+            notification.add(response.error, { type: "danger" });
+            return;
+        }
+        const threadId = upload.data.get("thread_id");
+        const threadModel = upload.data.get("thread_model");
+        const originThread =
+            messaging.state.threads[
+                threadModel === "mail.channel" ? parseInt(threadId) : `${threadModel},${threadId}`
+            ];
+        const attachment = {
```

Comment:
This is sad to leak technical detail of `thread.id` being channel id or a "local id". This adds technical debt to a bad design choice, which makes it slightly harder to change this bad design.